### PR TITLE
Porting to Poetry

### PR DESCRIPTION
**Pull request by DanielYakubov**
# Objective 

Hopefully avoid dependency hell forever with the addition of the lock file.

Just a note, this PR is likely going to have to go through a few phases while fighting with CI.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -29,16 +29,16 @@ jobs:
       - run:
           name: Build source distribution and install package from it
           command: |
-              pip install -r requirements.txt && \
+              poetry install && \
               python -m build && \
```
Comment (kylebgorman) [inline .circleci/config.yml:33]: is the `build` step stilll necessary then?

File: pyproject.toml [hunk 2]
```diff
@@ -44,9 +32,27 @@ classifiers = [
...
+exclude = ["configs*",  "examples*", "scripts*" ]
+
+[tool.poetry.dependencies]
+python = "^3.9"
+jsonargparse = { version = "^4.32.0", extras = ["signatures"] }
+lightning = "^2.4.0"
+transformers = "^4.44.0"
+torch = "^2.4.0"
+torchmetrics = "^1.4.0.post0"
+black = "^24.4.2"
```
Comment (kylebgorman) [inline pyproject.toml:45]: So a distinction is being lost here: this is not a dependency of the library at all and indeed users don't need it. But developers might. That's why it's in `requirements.txt` but not in `pyproject.toml` in the current design. Is there some way to encode that?

**Review by kylebgorman — COMMENTED** [review]
Quick comment...it looks like you got the CI testing running on your PRs.

File: pyproject.toml (hunk 2 above)
Comment (DanielYakubov) [reply pyproject.toml:50]: fixed

File: pyproject.toml [hunk 3]
```diff
@@ -44,9 +32,36 @@ classifiers = [
     "Operating System :: OS Independent",
     "Topic :: Text Processing :: Linguistic",
 ]
+include = ["udtube*"]
+exclude = ["configs*",  "examples*", "scripts*" ]
+
+[tool.poetry.dependencies]
+python = "^3.9"
```
Comment (kylebgorman) [inline pyproject.toml:39]: Right now we have Python 3.8 support listed elsewhere. It was end-of-lifed on October 7th though so we could drop it from `pyproject.toml`.

File: README.md [hunk 4]
```diff
@@ -37,7 +33,9 @@ also be specified using POSIX-style command-line flags.
 
 To install UDTube and its dependencies, run the following command:
 
-    pip install .
+    poetry install
+
+Note you'll have to have [poetry](https://python-poetry.org/docs/) installed
```
Comment (kylebgorman) [inline README.md:38]: Needs a period.

File: poetry.lock [hunk 5]
```diff
@@ -0,0 +1,3072 @@
+# This file is automatically @generated by Poetry 1.8.5 and should not be changed by hand.
+
+[[package]]
+name = "aiohappyeyeballs"
```
Comment (kylebgorman) [inline poetry.lock:4]: 👀

**Review by kylebgorman — APPROVED** [review]
LGTM, a few notes below.

File: pyproject.toml (hunk 3 above)
Comment (DanielYakubov) [reply pyproject.toml:39]: Okay, so I am understanding 3.8 is okay?

File: poetry.lock (hunk 5 above)
Comment (DanielYakubov) [reply poetry.lock:4]: Somehow it's a dependency, but the name is funny

File: pyproject.toml (hunk 3 above)
Comment (kylebgorman) [reply pyproject.toml:39]: yeah but change it (in two places) in `pyproject.toml` to eliminate references to 3.8 before submitting

Comment (DanielYakubov) [reply pyproject.toml:39]: Sorry, I don't see any references to 3.8 in the `pyproject.toml`, or elsewhere in the project. 3.9 support and up is present in this `pyproject.toml`. I see some in the poetry.lock, but that's autogenerated. Am I misunderstanding something?

Comment (kylebgorman) [reply pyproject.toml:39]: You're right, I was seeing an old cache of things, disregard.

**Pull request merged**